config/source.go

```diff
@@ -58,37 +58,59 @@ func ParseSource(cfg map[string]string) (Source, error) {
 		BatchSize:      defaultBatchSize,
 	}
 
-	keyColumns := strings.Split(cfg[KeyColumns], ",")
-	for i := range keyColumns {
-		if keyColumn := strings.TrimSpace(keyColumns[i]); keyColumn != "" {
+	if cfg[KeyColumns] != "" {
+		keyColumns := strings.Split(cfg[KeyColumns], ",")
+		for i := range keyColumns {
+			keyColumn := strings.TrimSpace(keyColumns[i])
+			if keyColumn == "" {
+				return Source{}, fmt.Errorf("invalid %q", KeyColumns)
+			}
+
 			sourceConfig.KeyColumns = append(sourceConfig.KeyColumns, keyColumn)
 		}
 	}
 
 	if cfg[Columns] != "" {
 		columnsSl := strings.Split(cfg[Columns], ",")
 		for i := range columnsSl {
-			if column := strings.TrimSpace(columnsSl[i]); column != "" {
-				sourceConfig.Columns = append(sourceConfig.Columns, column)
+			column := strings.TrimSpace(columnsSl[i])
+			if column == "" {
+				return Source{}, fmt.Errorf("invalid %q", Columns)
 			}
-		}
 
-		err = validateColumns(sourceConfig.OrderingColumn, sourceConfig.KeyColumns, sourceConfig.Columns)
-		if err != nil {
-			return Source{}, fmt.Errorf("validate config columns: %w", err)
+			sourceConfig.Columns = append(sourceConfig.Columns, column)
 		}
 	}
 
 	if cfg[BatchSize] != "" {
 		sourceConfig.BatchSize, err = strconv.Atoi(cfg[BatchSize])
 		if err != nil {
-			return Source{}, fmt.Errorf("parse BatchSize: %w", err)
+			return Source{}, fmt.Errorf("invalid %q: %w", BatchSize, err)
 		}
 	}
 
 	err = validate(sourceConfig)
 	if err != nil {
-		return Source{}, fmt.Errorf("validate source config: %w", err)
+		return Source{}, err
+	}
+
+	if len(sourceConfig.Columns) == 0 {
+		return sourceConfig, nil
+	}
+
+	columnsMap := make(map[string]struct{}, len(sourceConfig.Columns))
```

Comment:
maybe we can remove this variable and check key column and ordering column inside loop?

Reply:
We can check that `columns` contains the `orderingColumn` in a range of `columns`.
But I must check that the `columns` contains all `keyColumns` keys.